src/main/scala/org/apache/spark/mllib/linalg/distributed/BetterBlockMatrix.scala

```diff
@@ -0,0 +1,178 @@
+        val leftMat = leftBlock(row, i, context)
+        val rightMat = rightBlock(i, col, context)
+        (leftMat, rightMat) match {
+          case (x: DenseMatrix, y: DenseMatrix) =>
+            BLAS.gemm(1.0, x, y, 1.0, result)
+          case _ =>
+            throw new SparkException(s"No support for multiplying: ${leftMat.getClass} by ${rightMat.getClass}.")
+        }
+
+        i += 1
+      }
+
+      Iterator.single(((row, col), result))
+    }
+
+    val parts = (0 until nParts).map(IntPartition.apply _).toArray[Partition]
+    protected def getPartitions: Array[Partition] =
+      parts
+
+    /** Optionally overridden by subclasses to specify how they are partitioned. */
+    @transient override val partitioner: Option[Partitioner] =
+      Some(new GridPartitioner(rowBlocks, colBlocks, 1, 1))
+  }
+
+  case class IntPartition(index: Int) extends Partition { }
+
+  private def gridPartition(m: BlockMatrix): BlockMatrix = {
+    val p = new GridPartitioner(m.numRowBlocks, m.numColBlocks, 1, 1)
+    new BlockMatrix(m.blocks.partitionBy(p), m.rowsPerBlock, m.colsPerBlock, m.numRows(), m.numCols())
+  }
+
+  private def ensureGridPartitioning(m: BlockMatrix): BlockMatrix = m.blocks.partitioner match {
+    case Some(gp: GridPartitioner) if (gp.rowsPerPart == 1 && gp.colsPerPart == 1) =>
+      m
+    case Some(gp: GridPartitioner) =>
+      logDebug(s"Repartitioning a matrix (slow), $m, with a grid partitioner that didn't have 1 block per partition, had: ${gp.rowsPerPart} x ${gp.colsPerPart}")
+      gridPartition(m)
+    case Some(p) =>
+      logDebug(s"Repartitioning a matrix (slow), $m, with a non-grid partitioner: $p")
+      gridPartition(m)
+    case None =>
+      logDebug(s"Partitioning a matrix (slow), $m, with a no partitioner")
+      gridPartition(m)
+  }
+
+  def multiply(preL: BlockMatrix, preR: BlockMatrix): BlockMatrix = {
+    require(preL.numCols() == preR.numRows(),
+      s"""The number of columns of the left matrix and the number of rows of the right
+          matrix must be equal: ${preL.numRows()} x ${preL.numCols()}, ${preR.numRows()}
+          x ${preR.numCols()}. If you think they should be equal, try setting the
+          dimensions of A and B explicitly while initializing them.""".stripMargin)
+    // FIXME: don't require same blocksize on each matrix
+    require(preL.colsPerBlock == preR.rowsPerBlock,
+      s"""The number of columns in blocks of the left matrix and the number of rows in
+          blocks of the right matrix must be equal: ${preL.rowsPerBlock} x
+          ${preL.colsPerBlock}, ${preR.rowsPerBlock} x ${preR.colsPerBlock}. Generally,
+          all matrices should use square blocks of the same dimension.""".stripMargin)
+
+    val l = ensureGridPartitioning(preL)
+    val r = ensureGridPartitioning(preR)
+    new BlockMatrix(new BlockMatrixMultiplyRDD(l, r), l.rowsPerBlock, r.colsPerBlock, l.numRows(), r.numCols())
+  }
+
+  class BlockMatrixTransposeRDD(m: BlockMatrix)
```

Comment:
I thought we had a RDD that just changes (asserts) a partitioning, but it doesn't look like we do.  In light of my comments below, it seems like you could add it and write: `PartitionedRDD(m.blocks.map { ((i, j), m) => ((j, i), m.transpose) }, new GridPartitioner { ... })` but is also fine.